tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php

```diff
@@ -0,0 +1,209 @@
+
+    /**
+     * @var array<string, RestrictionRuleGroup[]>
+     */
+    private array $productRestrictionGroupsByReference = [];
+
+    /**
+     * This is just a random number which in theory should never be reached as cart rule id in tests
+     */
+    private const NON_EXISTING_CART_RULE_ID = 54440051;
+
+    /**
+     * @When I restrict following cart rules for cart rule :cartRuleReference:
+     *
+     * @param string $cartRuleReference
+     * @param TableNode $tableNode
+     *
+     * @return void
+     */
+    public function setRestrictedCartRules(string $cartRuleReference, TableNode $tableNode): void
+    {
+        $command = $this->getRestrictionsCommand($cartRuleReference);
+
+        try {
+            $command->setRestrictedCartRuleIds($this->referencesToIds($tableNode->getRowsHash()['restricted cart rules']));
+            $this->restrictionCommandsByReference[$cartRuleReference] = $command;
+        } catch (CartRuleConstraintException $e) {
+            $this->setLastException($e);
+        }
+    }
+
+    /**
+     * @When I add a restriction for cart rule :cartRuleReference, which requires at least :quantity product(s) in cart matching one of these rules:
+     * @When I add a restriction for cart rule :cartRuleReference, which requires any quantity of product(s) in cart matching one of these rules:
+     * @When I add a restriction for cart rule :cartRuleReference, which requires at least :quantity product(s), but I provide empty list of rules
+     *
+     * @param string $cartRuleReference
+     * @param int $quantity
+     * @param TableNode|null $table
+     *
+     * @return void
+     */
+    public function addRestrictionRule(string $cartRuleReference, int $quantity = 0, ?TableNode $table = null): void
+    {
+        $rules = [];
+
+        try {
+            if ($table) {
+                foreach ($table->getColumnsHash() as $row) {
+                    $rules[] = new RestrictionRule($row['type'], $this->referencesToIds($row['references']));
+                }
+            }
+
+            $restrictionGroups = [];
+            if (isset($this->productRestrictionGroupsByReference[$cartRuleReference])) {
+                $restrictionGroups = $this->productRestrictionGroupsByReference[$cartRuleReference];
+            }
+
+            $restrictionGroups[] = new RestrictionRuleGroup($quantity, $rules);
+            $this->productRestrictionGroupsByReference[$cartRuleReference] = $restrictionGroups;
+        } catch (CartRuleConstraintException $e) {
+            $this->setLastException($e);
+        }
+    }
+
+    /**
+     * @When I restrict cart rules for :cartRuleReference providing non-existing cart rules
+     *
+     * @param string $cartRuleReference
+     *
+     * @return void
+     */
+    public function restrictCartRulesProvidingNonExistingIds(string $cartRuleReference): void
+    {
+        $command = $this->getRestrictionsCommand($cartRuleReference);
```

Comment:
In this endpoint, you execute the command right away so you shouldn't rely on the tmp saved command here because it could already contain some values inconsistent ith what the endpoint aims at doing

I think it would be safer to create a new dedicated command, set the invalid value and execute it, additionally, the persisted command could even be cleared.

TLDR; `getRestrictionsCommand` should ONLY be used in the endpoints that need stateful steps because of their complexity, this one can remain independent and stateless